File: go-chi/app/outgoing/taskrepo/postgres.go
package taskrepo

import (
	"github.com/google/uuid"
	"github.com/jmoiron/sqlx"
)

type Postgres struct {
	database *sqlx.DB
}

func NewPostgres(database *sqlx.DB) *Postgres {
	return &Postgres{
		database: database,
	}
}

func (taskRepository *Postgres) FindAll() ([]TaskEntity, error) {
	var taskEntities []TaskEntity
	err := taskRepository.database.Select(&taskEntities, "SELECT task_id, title, description FROM public.tasks")
	if err != nil {
		return nil, err
	}
	return taskEntities, nil
}

func (taskRepository *Postgres) FindByTaskId(taskId uuid.UUID) (TaskEntity, error) {
	var taskEntity TaskEntity
	err := taskRepository.database.Get(&taskEntity, "SELECT task_id, title, description FROM public.tasks WHERE task_id = $1", taskId)
	if err != nil {
		return TaskEntity{}, err
	}
	return taskEntity, nil
}

func (taskRepository *Postgres) Save(taskEntity TaskEntity) (TaskEntity, error) {
	_, err := taskRepository.database.NamedExec("INSERT INTO public.tasks (task_id, title, description) VALUES (:task_id, :title, :description)", taskEntity)
	if err != nil {
		return TaskEntity{}, err
	}
	return taskEntity, nil
}

func (taskRepository *Postgres) Update(taskEntity TaskEntity) error {
	result, err := taskRepository.database.NamedExec(`
		UPDATE public.tasks
		SET title = :title, description = :description
		WHERE task_id = :task_id`,
		taskEntity,
	)
	if err != nil {
		return err
	}

	affectedRows, err := result.RowsAffected()
	if err != nil {
		return err
	}
	if affectedRows == 0 {
		return ErrTaskNotFound
	}

	return nil
}

func (taskRepository *Postgres) DeleteByTaskId(taskId uuid.UUID) error {
	result, err := taskRepository.database.Exec("DELETE FROM public.tasks WHERE task_id = $1", taskId)
	if err != nil {
		return err
	}

	affectedRows, err := result.RowsAffected()
	if err != nil {
		return err
	}
	if affectedRows == 0 {
		return ErrTaskNotFound
	}

	return nil

}

File: rust-axum/app/src/outgoing/postgres.rs
use crate::outgoing::entities::TaskEntity;
use crate::outgoing::task_repo::TaskRepository;
use async_trait::async_trait;

use uuid::Uuid;

pub struct PostgresTaskRepository {
    postgres_pool: sqlx::PgPool,
}

impl PostgresTaskRepository {
    pub fn new(postgres_pool: sqlx::PgPool) -> PostgresTaskRepository {
        return PostgresTaskRepository {
            postgres_pool: postgres_pool,
        };
    }
}

#[async_trait]
impl TaskRepository for PostgresTaskRepository {
    async fn find_all(&self) -> Vec<TaskEntity> {
        return sqlx::query_as::<_, TaskEntity>("SELECT task_id, title, description FROM public.tasks")
            .fetch_all(&self.postgres_pool)
            .await
            .unwrap();
    }

    async fn find_by_task_id(&self, task_id: Uuid) -> Option<TaskEntity> {
        return sqlx::query_as::<_, TaskEntity>(
            "SELECT task_id, title, description FROM public.tasks WHERE task_id = $1",
        )
        .bind(task_id)
        .fetch_optional(&self.postgres_pool)
        .await
        .unwrap();
    }

    async fn save(&mut self, task: TaskEntity) -> TaskEntity {
        return sqlx::query_as::<_, TaskEntity>(
            "INSERT INTO public.tasks (task_id, title, description) VALUES ($1, $2, $3) RETURNING task_id, title, description",
        )
        .bind(task.task_id)
        .bind(task.title)
        .bind(task.description)
        .fetch_one(&self.postgres_pool)
        .await
        .unwrap();
    }

    async fn update(&mut self, task: TaskEntity) -> Option<TaskEntity> {
        return sqlx::query_as::<_, TaskEntity>(
            "UPDATE public.tasks SET title = $1, description = $2 WHERE task_id = $3 RETURNING task_id, title, description",
        )
        .bind(task.title)
        .bind(task.description)
        .bind(task.task_id)
        .fetch_one(&self.postgres_pool)
        .await
        .ok();
    }

    async fn delete_by_task_id(&mut self, task_id: Uuid) -> Option<TaskEntity> {
        return sqlx::query_as::<_, TaskEntity>(
            "DELETE FROM public.tasks WHERE task_id = $1 RETURNING task_id, title, description",
        )
        .bind(task_id)
        .fetch_optional(&self.postgres_pool)
        .await
        .unwrap();
    }
}
